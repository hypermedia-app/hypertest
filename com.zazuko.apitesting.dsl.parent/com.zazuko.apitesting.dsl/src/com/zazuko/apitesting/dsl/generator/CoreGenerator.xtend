/*
 * generated by Xtext 2.17.0
 */
package com.zazuko.apitesting.dsl.generator

import com.zazuko.apitesting.dsl.core.ClassBlock
import com.zazuko.apitesting.dsl.core.ClassLevelAssertion
import com.zazuko.apitesting.dsl.core.PropertyAssertion
import org.eclipse.emf.ecore.resource.Resource
import org.eclipse.xtext.generator.AbstractGenerator
import org.eclipse.xtext.generator.IFileSystemAccess2
import org.eclipse.xtext.generator.IGeneratorContext

/**
 * Generates code from your model files on save.
 * 
 * See https://www.eclipse.org/Xtext/documentation/303_runtime_concepts.html#code-generation
 */
class CoreGenerator extends AbstractGenerator {

	override void doGenerate(Resource resource, IFileSystemAccess2 fsa, IGeneratorContext context) {
		
		val Iterable<ClassBlock> blocks = resource.allContents.filter(ClassBlock).toList
		
		if ( ! blocks.empty) {
			val String dslFileName = resource.getURI().lastSegment.toString();
			val String outFileBase = dslFileName.substring(0, dslFileName.lastIndexOf("."));
			
			fsa.generateFile(outFileBase + '.json', generateSteps(blocks));
		}		
	}	
	
	def generateSteps(Iterable<ClassBlock> blocks) '''
		{
			"steps": [
				«FOR block:blocks SEPARATOR ","»
					«block.step»
				«ENDFOR»
			]
		}
	'''
	
	def step(ClassBlock it) '''
		{
			"type": "Class",
			"classId": "«name»",
			"children": [
				«FOR assertion:assertions SEPARATOR ","»
					«assertion.child»
				«ENDFOR»
			]
			«hatch»
		}
	'''	
	
	def dispatch CharSequence child(PropertyAssertion it)  '''
		{
			"type": "Property",
			"propertyId": "«name»",
			"children": [
				«FOR assertion:assertions SEPARATOR ","»
					«assertion.child»
				«ENDFOR»
			]
			«hatch»
		}
	'''
	def dispatch child(ClassLevelAssertion it) '''
		# TODO: implementation missing for child(«class.name»)
	'''
	
}