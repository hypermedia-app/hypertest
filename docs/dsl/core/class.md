# Class Step

While a `Class` is not a canonical element of hypermedia, some of the more feature-rich
media types, such as Kevin Swiber's [SIREN](https://github.com/kevinswiber/siren) and
[Hydra](http://www.hydra-cg.com/spec/latest/core/) offer an additional piece of information to quickly
tell one resource representation from another by introducing `class` metadata.

## Remarks

A class is only valid at top-level.

## Example

<<< @/api-examples/core/class/empty-class.api
