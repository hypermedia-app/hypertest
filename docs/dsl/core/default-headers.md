# Scenario Headers Block

At the top of a hypertest scenario a `HEADERS` block may be defined which defines HTTP headers
added to every request within this scenario.

This is useful for example to pass authentication information.

## Remarks

Any field name can used multiple times even if the given header's semantics does not allow it. It is up to
the test runner to correctly set them on the server request.

## Example

<<< @/api-examples/core/default-headers/auth-and-cookies.api
