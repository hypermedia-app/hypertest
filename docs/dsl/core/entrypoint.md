# Entrypoint Statement

It may be desired to start the test scenario by requesting

## Remarks

There can be only one `ENTRYPOINT` statement at the top of a scenario document.

The test runner should apply it relative to the base URL provided as the command parameter.

## Example

When running the below against an API under `http://tech.blog/api/`, the first requested resource would be
`http://tech.blog/api/some/article`.

<<< @/api-examples/core/entrypoint/custom-entry.api
