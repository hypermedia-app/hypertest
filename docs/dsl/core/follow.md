# Follow Step

Combined with `Expect Location`, the follow step can be used to request the target
URL which was the result of a create response.

## Usage

### Location header

The `Location` header is typically set on a `201 Created` response. The example below
shows how such a response from an [`Operation`] can be handled by following the new 
resource's URI set on the response:

<<< @/api-examples/hydra/follow/location-header.hydra
