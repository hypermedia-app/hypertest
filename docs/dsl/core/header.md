# Header step

Header step is a way to inspect [HTTP headers][hh] in a generic way.

[hh]: https://developer.mozilla.org/docs/Web/HTTP/Headers

## Syntax

* Header statement is only valid where response is the context
* The header name is case-insensitive, as the HTTP header fields are

## Usage

### Check link exists on the response

<<< @/api-examples/core/header/exists.api

### Text header value

To check that a header contains a specific value it is possible to do an exact
comparison or regular expression matching.

<<< @/api-examples/core/header/assert-value.api

:::tip
The regular expression string is not parsed as the exact implementation would
depend on the test runner.
:::

### Capture header value

It is also possible to create a "variable" containing the header's value, so that
it can be used in subsequent steps.

<<< @/api-examples/core/header/capture-value.api
