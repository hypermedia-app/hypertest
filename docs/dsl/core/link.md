# Link Step

Link steps let test scenarios traverse between resource representations in order to
analyse those linked resources.

:::tip
It is up to the runner implementation to decide whether the link step should
only apply to links found in bodies (such as [HAL's][hal] `_links` or
HTML's [anchor element (`<a>`)][html-a] or also link headers.

Alternatively HTTP link headers can also be handled with [`Expect Header` statement](./header).
:::

[html-a]: https://developer.mozilla.org/docs/Web/HTML/Element/a
[hal]: http://stateless.co/hal_specification.html

## Syntax

* Link steps are valid where representation is the context.
* Link steps are valid on top level.

## Usage

### Top-level link

Link block can be placed at the top of a test scenario document. This way it such a
link will be followed from any resource where it is found.

<<< @/api-examples/core/link/top-level.api

### Resource-specific links

Links can be used anywhere that resource representation is the block context. For example,
in [`Class block`](./class.md) or [`Property block`](./property.md).

<<< @/api-examples/core/link/nested.api

### Strict check

The example above can be modified to require that the presence of a link.

<<< @/api-examples/core/link/nested-mandatory.api

### Usage without children

A strict `Expect Link` can also be a statement, thus only checking that the link exists,
without performing any specific child checks.

<<< @/api-examples/core/link/statement.api

:::tip
The will should be dereferenced regardless and any applicable top-level steps
will be performed.
:::
