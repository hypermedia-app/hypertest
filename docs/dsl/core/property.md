# Property Step

A property step lets the test author inspect properties or attributes of resource
representations.

## Syntax

* Property steps are valid where representation is the context.
* Property steps are not valid on top level.

## Usage

### Strict property check

A test may require that a given property is present. If not, then a test failure should be
reported.

```
Expect Property "name"
```

### Asserting literal property value

An optional parameter can be provided, which instructs the client to compare it with 
the actual value of the property.

```
Expect Property "name" "Tomasz"
```

:::warning
At the time of writing only string is allowed. The test runner should always compare
textual form of non-string values.
:::

### Step into nested resource

A property step can also be a block.

If a property is not necessary for the success of the test scenario, it may be treated as 
optional, by using `With` modifier instead of `Expect` modifier.

```
With Class "Book" {
    
    # Book may have an embedded author resource
    # Use Expect Property if author is manadatory
    With Property "author" {
    
        # If it does't check that it has a name
        Expect Property "name"
    }
}
```
