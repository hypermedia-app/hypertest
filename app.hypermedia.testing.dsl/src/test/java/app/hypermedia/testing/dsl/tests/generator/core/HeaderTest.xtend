package app.hypermedia.testing.dsl.tests.generator.core

import org.eclipse.xtext.generator.InMemoryFileSystemAccess
import org.junit.jupiter.api.^extension.ExtendWith
import org.eclipse.xtext.testing.extensions.InjectionExtension
import org.eclipse.xtext.testing.InjectWith
import com.google.inject.Inject
import org.eclipse.xtext.testing.util.ParseHelper
import app.hypermedia.testing.dsl.core.CoreScenario
import org.junit.jupiter.api.Test
import static io.github.jsonSnapshot.SnapshotMatcher.*
import org.eclipse.xtext.generator.IGenerator2
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.AfterAll
import org.json.JSONObject
import app.hypermedia.testing.dsl.tests.CoreInjectorProvider
import org.eclipse.xtext.generator.GeneratorContext

@ExtendWith(InjectionExtension)
@InjectWith(CoreInjectorProvider)
class HeaderTest {
    @Inject IGenerator2 generator
    @Inject extension ParseHelper<CoreScenario>
    @BeforeAll
    static def beforeAll() {
        start()
    }

    @AfterAll
    static def afterAll() {
        validateSnapshots();
    }

    @Test
    def emptyHeaderNameOnly_generatesStep() {
        // given
        val model = '''
            With Link "author" {
                Expect Header ETag
            }
        '''.parse

        // when
        val fsa = new InMemoryFileSystemAccess()
        generator.doGenerate(model.eResource, fsa, new GeneratorContext())
        println(fsa.textFiles)

        // then
        val file = new JSONObject(fsa.textFiles.values.get(0).toString)
        expect(file).toMatchSnapshot()
    }

    @Test
    def headerExactValueAssertion_generatesStep() {
        // given
        val model = '''
            With Link "author" {
                Expect Header Content-Type "application/problem+json"
            }
        '''.parse

        // when
        val fsa = new InMemoryFileSystemAccess()
        generator.doGenerate(model.eResource, fsa, new GeneratorContext())
        println(fsa.textFiles)

        // then
        val file = new JSONObject(fsa.textFiles.values.get(0).toString)
        expect(file).toMatchSnapshot()
    }

    @Test
    def headerRegexValueAssertion_generatesStep() {
        // given
        val model = '''
            With Link "author" {
                Expect Header Content-Type Matching "^application/"
            }
        '''.parse

        // when
        val fsa = new InMemoryFileSystemAccess()
        generator.doGenerate(model.eResource, fsa, new GeneratorContext())
        println(fsa.textFiles)

        // then
        val file = new JSONObject(fsa.textFiles.values.get(0).toString)
        expect(file).toMatchSnapshot()
    }

    @Test
    def headerWithVariableCapture_generatesStep() {
        // given
        val model = '''
            With Link "author" {
                Expect Header Location [uri]
            }
        '''.parse

        // when
        val fsa = new InMemoryFileSystemAccess()
        generator.doGenerate(model.eResource, fsa, new GeneratorContext())
        println(fsa.textFiles)

        // then
        val file = new JSONObject(fsa.textFiles.values.get(0).toString)
        expect(file).toMatchSnapshot()
    }
}
