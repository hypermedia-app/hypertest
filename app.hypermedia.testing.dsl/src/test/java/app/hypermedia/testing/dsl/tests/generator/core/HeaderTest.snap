app.hypermedia.testing.dsl.tests.generator.core.HeaderTest.emptyHeaderNameOnly_generatesStep=[
  {
    "map": {
      "steps": {
        "myArrayList": [
          {
            "map": {
              "children": {
                "myArrayList": [
                  {
                    "map": {
                      "header": "ETag",
                      "type": "ResponseHeader"
                    },
                    "empty": false
                  }
                ],
                "empty": false
              },
              "rel": "author",
              "strict": false,
              "type": "Link"
            },
            "empty": false
          }
        ],
        "empty": false
      }
    },
    "empty": false
  }
]


app.hypermedia.testing.dsl.tests.generator.core.HeaderTest.headerExactValueAssertion_generatesStep=[
  {
    "map": {
      "steps": {
        "myArrayList": [
          {
            "map": {
              "children": {
                "myArrayList": [
                  {
                    "map": {
                      "header": "Content-Type",
                      "type": "ResponseHeader",
                      "value": "application/problem+json"
                    },
                    "empty": false
                  }
                ],
                "empty": false
              },
              "rel": "author",
              "strict": false,
              "type": "Link"
            },
            "empty": false
          }
        ],
        "empty": false
      }
    },
    "empty": false
  }
]


app.hypermedia.testing.dsl.tests.generator.core.HeaderTest.headerRegexValueAssertion_generatesStep=[
  {
    "map": {
      "steps": {
        "myArrayList": [
          {
            "map": {
              "children": {
                "myArrayList": [
                  {
                    "map": {
                      "header": "Content-Type",
                      "pattern": "^application/",
                      "type": "ResponseHeader"
                    },
                    "empty": false
                  }
                ],
                "empty": false
              },
              "rel": "author",
              "strict": false,
              "type": "Link"
            },
            "empty": false
          }
        ],
        "empty": false
      }
    },
    "empty": false
  }
]


app.hypermedia.testing.dsl.tests.generator.core.HeaderTest.headerWithVariableCapture_generatesStep=[
  {
    "map": {
      "steps": {
        "myArrayList": [
          {
            "map": {
              "children": {
                "myArrayList": [
                  {
                    "map": {
                      "captureAs": "uri",
                      "header": "Location",
                      "type": "ResponseHeader"
                    },
                    "empty": false
                  }
                ],
                "empty": false
              },
              "rel": "author",
              "strict": false,
              "type": "Link"
            },
            "empty": false
          }
        ],
        "empty": false
      }
    },
    "empty": false
  }
]